client/go/outline/reporting/reporting_client.go

```diff
@@ -0,0 +1,274 @@
+		return fmt.Errorf("failed to write request: %w", err)
+	}
+
+	// Read the response from the server.
+	var response []byte
+	buffer := make([]byte, 4096)
+	for {
+		n, err := conn.Read(buffer)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("failed to read response: %w", err)
+		}
+		response = append(response, buffer[:n]...)
+	}
+
+	// Parse the HTTP response to extract cookies.
+	headersEnd := bytes.Index(response, []byte("\r\n\r\n"))
+	if headersEnd == -1 {
+		return fmt.Errorf("failed to find end of headers in response")
+	}
+
+	headers := string(response[:headersEnd])
+	fmt.Println("HTTP Headers:")
+	fmt.Println(headers)
+
+	// Extract cookies from the headers.
+	var newCookies []*http.Cookie
+	lines := strings.Split(headers, "\r\n")
+	for _, line := range lines {
+		if strings.HasPrefix(line, "Set-Cookie:") {
+			cookie := strings.TrimPrefix(line, "Set-Cookie: ")
+			parts := strings.SplitN(cookie, ";", 2)
+			if len(parts) > 0 {
+				cookieParts := strings.SplitN(parts[0], "=", 2)
+				if len(cookieParts) == 2 {
+					newCookies = append(newCookies, &http.Cookie{
+						Name:  cookieParts[0],
+						Value: cookieParts[1],
+					})
+					fmt.Println("Cookie found:", cookie)
+				}
+			}
+		}
+	}
+	// Check if the URL is already in the list
+	exists := false
+	for _, existingURL := range urls {
+		if existingURL.String() == u.String() {
+			exists = true
+			break
+		}
+	}
+	if !exists {
+		urls = append(urls, u)
+	}
+	jar.SetCookies(u, newCookies)
+	// Save new cookies to file
+	if err := saveCookies(urls, jar, cookiesFile); err != nil {
+		return fmt.Errorf("failed to save cookies: %w", err)
+	}
+
+	return nil
+}
+
+// CookieData represents the structure for saving cookies, including the URL
+type CookieData struct {
+	URL     string         `json:"url"`
+	Cookies []CookieDetail `json:"cookies"`
+}
+
+type CookieDetail struct {
+	Name   string `json:"name"`
+	Value  string `json:"value"`
+	Domain string `json:"domain"`
+	Path   string `json:"path"`
+}
+
+func saveCookies(urls []*url.URL, jar http.CookieJar, filename string) error {
+	// Retrieve all cookies from the jar
+	cookieDataMap := make(map[string][]CookieDetail)
+
+	for _, urlObj := range urls {
+		cookies := jar.Cookies(urlObj)
+		var cookieDetails []CookieDetail
+		for _, cookie := range cookies {
+			cookieDetails = append(cookieDetails, CookieDetail{
+				Name:   cookie.Name,
+				Value:  cookie.Value,
+				Domain: cookie.Domain,
+				Path:   cookie.Path,
+			})
+		}
+
+		cookieDataMap[urlObj.String()] = cookieDetails
+	}
+
+	// Save the cookies data to a file
+	file, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	encoder := json.NewEncoder(file)
+	if err := encoder.Encode(cookieDataMap); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func loadCookies(jar http.CookieJar, filename string) ([]*url.URL, error) {
+	// Read cookies from the file
+	file, err := os.Open(filename)
+	if os.IsNotExist(err) {
```

Comment:
According to [Go doc](https://pkg.go.dev/os#IsNotExist):

> This function predates [errors.Is](https://pkg.go.dev/errors#Is). It only supports errors returned by the os package. New code should use errors.Is(err, fs.ErrNotExist).

```suggestion
	if errors.Is(err, fs.ErrNotExist) {
```